shell/src/main/java/alluxio/cli/fs/command/CpCommand.java

```diff
@@ -104,29 +117,42 @@
      * NOTE: needs to call {@link #shutdown()} to release resources.
      *
      * @param threads number of threads
-     * @param out the output stream for tasks to send messages to
+     * @param stdout the stdout stream for tasks to send messages to
+     * @param stderr the stderr stream for tasks to send error messages to
+     * @param fileSystem the Alluxio filesystem used to delete path
      * @param path the path to delete on shutdown when it's empty, otherwise can be {@code null}
      */
-    public CopyThreadPoolExecutor(int threads, PrintStream out, AlluxioURI path) {
+    public CopyThreadPoolExecutor(int threads, PrintStream stdout, PrintStream stderr,
+        FileSystem fileSystem, AlluxioURI path) {
       mPool = new ThreadPoolExecutor(threads, threads,
           1, TimeUnit.SECONDS, new ArrayBlockingQueue<>(threads * 2),
           new ThreadPoolExecutor.CallerRunsPolicy());
       mMessages = new LinkedBlockingQueue<>();
-      mOutput = out;
+      mExceptions = new ConcurrentLinkedQueue<>();
+      mStdout = stdout;
+      mStderr = stderr;
       mPrinter = new Thread(() -> {
         while (!Thread.currentThread().isInterrupted()) {
           try {
-            String message = mMessages.take();
-            if (message.equals(MESSAGE_DONE)) {
+            Object message = mMessages.take();
+            if (message == MESSAGE_DONE) {
               break;
             }
-            mOutput.println(message);
+            if (message instanceof String) {
+              mStdout.println(message);
+            } else if (message instanceof Exception) {
+              mStderr.println(((Exception) message).getMessage());
```

Comment:
nit: does the exception always include the path for which the exception occured? if it is not guaranteed, I would format the output as "Failed to copy path {} with exception {}"

Reply:
I refactored the `sendMessage` API so that the messages are in a better controlled format.